java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -196,8 +194,27 @@ public void visitContinueStatement(ContinueStatementTree tree) {
   @Override
   public void visitMethodInvocation(MethodInvocationTree tree) {
     MethodInvocationTreeImpl mit = (MethodInvocationTreeImpl) tree;
-    if(mit.isTypeSet()) {
-      // nothing special to deduce once we have set type of expression
+    if(mit.isTypeSet() && mit.symbol().isMethodSymbol()) {
+      JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) mit.symbol();
+      JavaType parametrizedTypeType = (JavaType) mit.symbolType();
+      if(parametrizedTypeType.isTagged(JavaType.PARAMETERIZED)) {
+        TypeSubstitution typeSubstitution = ((ParametrizedTypeJavaType) parametrizedTypeType).typeSubstitution.replaceKeys(methodSymbol.typeVariableTypes);
+        parametrizedTypeType = parametrizedTypeCache.getParametrizedTypeType((JavaSymbol.TypeJavaSymbol) parametrizedTypeType.symbol(), typeSubstitution);
+      }
+      final JavaType finalParametrizedTypeType = parametrizedTypeType;
+      List<JavaType> inferedArgTypes =
+        methodSymbol.parameterTypes().stream().map(formalType -> resolve.resolveTypeSubstitution((JavaType) formalType, finalParametrizedTypeType)).collect(Collectors.toList());
+      List<JavaType> argTypes = getParameterTypes(tree.arguments());
+      for (int i = 0; i < argTypes.size(); i++) {
+        JavaType arg = argTypes.get(i);
+        int size = inferedArgTypes.size();
+        Type formal = inferedArgTypes.get((i < size) ? i : (size - 1));
+        if (formal != arg) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)